counterpartylib/lib/transaction.py

```diff
@@ -451,111 +444,109 @@ def construct (db, tx_info, encoding='auto',
-        # Approximate needed change, fee by with most recently calculated
-        # quantities.
-        btc_out = destination_btc_out + data_btc_out
-        total_btc_out = btc_out + max(change_quantity, 0) + final_fee
-        raise exceptions.BalanceError('Insufficient {} at address {}. (Need approximately {} {}.) To spend unconfirmed coins, use the flag `--unconfirmed`. (Unconfirmed coins cannot be spent from multi‐sig addresses.)'.format(config.BTC, source, total_btc_out / config.UNIT, config.BTC))
-
-    # Lock the source's inputs (UTXOs) chosen for this transaction
-    if UTXO_LOCKS is not None and not disable_utxo_locks:
-        if source not in UTXO_LOCKS:
-            UTXO_LOCKS[source] = cachetools.TTLCache(
-                UTXO_LOCKS_PER_ADDRESS_MAXSIZE, config.UTXO_LOCKS_MAX_AGE)
-
-        for input in inputs:
-            UTXO_LOCKS[source][make_outkey(input)] = input
-
-        logger.debug("UTXO locks: Potentials ({}): {}, Used: {}, locked UTXOs: {}".format(
-            len(unspent), [make_outkey(coin) for coin in unspent],
-            [make_outkey(input) for input in inputs], list(UTXO_LOCKS[source].keys())))
-
-    '''Finish'''
-
-    # Change output.
-    if change_quantity:
-        if script.is_multisig(source):
-            change_address = backend.multisig_pubkeyhashes_to_pubkeys(source, provided_pubkeys)
+            destination_value_sum = sum([value for (destination, value) in destination_outputs])
+            source_value = destination_value_sum
+
+            logger.getChild('p2shdebug').debug('source_value %d' % (source_value, ))
+            logger.getChild('p2shdebug').debug('change_value %d' % (change_output[1], ))
+            if change_output:
+                # add the difference between source and destination to the change
+                change_value = change_output[1] + (destination_value_sum - source_value)
+                change_output = (change_output[0], change_value)
+
+            unsigned_pretx = serializer.serialise_p2sh_pretx(inputs,
+                                                             source=source_address,
+                                                             source_value=source_value,
+                                                             data_output=data_output,
+                                                             change_output=change_output,
+                                                             pubkey=dust_return_pubkey,
+                                                             multisig_pubkeys=p2sh_source_multisig_pubkeys,
+                                                             multisig_pubkeys_required=p2sh_source_multisig_pubkeys_required)
+            unsigned_pretx_hex = binascii.hexlify(unsigned_pretx).decode('utf-8')
+
+        # with segwit we already know the txid and can return both
+        if segwit:
+            pretx_txid = hashlib.sha256(unsigned_pretx).digest()  # this should be segwit txid
```

Comment:
this isn't correct cuz we stuff the scriptPubKey into the scriptSig so the signing party knows what to sign with, I think we should just remove the segwit stuff for now until counterwallet actually supports it maybe?